pkg: u8glib: add initial support for graphics library

basilfx opened the pull request:
This PR adds initial support for [U8glib](https://github.com/olikraus/u8glib) graphics library as an external package. While targeted for Arduino MCUs, it works fine on 32-bit MCUs. The license should be compatible (new-style BSD).

I've tested this with a cheap [SSD1306 OLED display](https://www.youtube.com/watch?v=bG5_EwFTiDc) via SPI, but it also has a stdout (for native) and I2C backend (untested). The included test application cycles through three pages, each second. Several [other](https://github.com/olikraus/u8glib/wiki/device) displays are also supported.

![u8glib](https://cloud.githubusercontent.com/assets/815976/13935668/cf1a3880-efb8-11e5-9698-bc660416f5a4.jpg)

The code size for the test application only (EFM32, STK3600, SPI) is approximately 4.0 kb.

OlegHahm commented:
nice!

OlegHahm commented on pkg/u8glib/Makefile:
```diff
@@ -0,0 +1,36 @@
+	cd "$(PKG_DIR)" && git am --ignore-whitespace "$(CURDIR)"/*.patch
+
+$(PKG_DIR)/.git/config:
+	test -d "$(PKG_DIR)" || git clone "$(PKG_URL)" "$(PKG_DIR)"; \
```
Can you update this Makefile to use "git cache" from #5112, please?

basilfx commented:
Added git-cache support and fixed some gcc errors (I normally use clang).

aabadie commented:
Thanks for the PR ! That was something I was also thinking about.
I tested it yesterday with a samr21-xpro and the [OLED1 Xplained extension](http://www.atmel.com/images/atmel-42077-oled1-xplained-pro_user-guide.pdf). The extension uses SPI so, according to the datasheet, I adapted the test application using this SPI initialization call:
```c
#define SPI_INTERFACE SPI_1
#define SS_PIN GPIO_PIN(PB,03)
#define A0_PIN GPIO_PIN(PA,13)
#define RS_PIN GPIO_PIN(PA,23)

int main(void)
{
    uint32_t screen = 0;
    u8g_t u8g; 
    u8g_spi_t dev;
    
    if (u8g_init_spi(&dev, SPI_INTERFACE, SS_PIN, A0_PIN, RS_PIN, &u8g, &u8g_dev_ssd1306_128x32_hw_spi)) {
        puts("Failed to initialize U8glib to SPI device.");
        return -1;
    }

/* draw text */
...
```

Unfortunately, nothing is displayed. I plan to test with another board asap (arduino-due for instance).

basilfx commented:
> Unfortunately, nothing is displayed. I plan to test with another board asap (arduino-due for instance).

Hmm, not sure what is wrong. You can try to swap SS/A0/RS pins and see if that makes a difference. This is because my cheap OLED display uses vague names, so I basically guessed and tried until it worked :-)

You can also try to set breakpoints in the `u8g_com_riotos_hw_spi_fn` method and see if the correct pins toggle.

Does the native/stdout backend work for you?

OlegHahm commented:
I tested the native/stdout and it worked for me. Don't have any other hardware to test right now.

basilfx commented:
Adapted to new pkg structure and rebased.

basilfx commented:
I bought the same SSD1306 OLED display, but this time with an I2C interface. I have updated this PR so that both interfaces work.

![bestand 02-05-16 22 25 43](https://cloud.githubusercontent.com/assets/815976/14966911/f6485028-10b4-11e6-90ca-dcbbf1f8c84b.jpeg)

(The test application only supports one display at a time, but the graphics memory persists.)

basilfx commented:
Fixed couple of things to make Murdock happy:

* Removed interface drivers from U8glib because of conflicts with the Arduino and MSP430 builds. These builds have their own defines (e.g. `ARDUINO`, `__MSP430__`) which would compile different code that would circumvent the RIOT-OS interface.
* Unified file names for RIOT-OS interface

Now Murdock is happy.

kaspar030 commented on tests/pkg_u8glib/main.c:
```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (C) 2015 Freie Universität Berlin
```
Is this the copyright you want to add? It's 2016. ;)

kaspar030 commented:
Please address copyright if needed, then squash.
We usually squash into one commit for the package, one for the test.

Very nice, I wish I had such a display!

basilfx commented:
Changed the copyright and squashed into two separate commits.

basilfx commented:
After #5435, this PR should build again.

basilfx commented:
Don't merge this PR yet. I'm improving the interface to RIOT-OS, which depends on a [PR](https://github.com/olikraus/u8glib/pull/411) I have submitted to U8glib. I hope that gets merged, otherwise I will point it to my own branch.

In the mean time, enjoy this picture of this PR running on a low-power memory display :-)

![result](https://cloud.githubusercontent.com/assets/815976/15445594/e5183362-1f01-11e6-8e18-00ab28568dbe.jpg)

basilfx commented:
I will close this PR as I am working on porting the successor of U8glib to RIOT-OS. The successor is (IMHO) much easier to use, includes the fixes I originally PR'd for U8glib and is less coupled to Arduino. It also need less patches.

I'll open a new PR once it is ready.

miri64 commented:
It would have been perfectly fine to just rework this PR, but no problem. Looking forward for the new PR :-).
